### PR TITLE
Fix new mods and sockets in the optimiser

### LINKED ISSUE #6192: The loadout optimizer can't find any loadout while any season mod and exotic armor were selected at the same time
**What version of DIM are you using?**
Version 6.39.1.9108 (beta), built on 2020/11/18

**What Browser and OS are you using?**
Edge(chrome base)

**Describe the bug and how to make it happen**
The loadout optimizer can't find any loadout while any season mod and exotic armor were selected at the same time.

### LINKED ISSUE #6195: Seasonal mods can only be assigned to last wish mod slots
**What version of DIM are you using?**
Version 6.39.1 (release), built on 11/16/2020, 11:58:15 AM
Also replicated on Version 6.39.1.9108 (beta), built on 11/17/2020, 11:48:25 PM

**What Browser and OS are you using?**
Chrome Version 87.0.4280.66

**Describe the bug and how to make it happen**
Loadout optimizer shows "No builds found" when selecting a seasonal mod and excluding all last wish gear. Possibly related to https://github.com/DestinyItemManager/DIM/issues/6191?

Here I've selected global reach and left 1 last wish gauntlet un-excluded.
last_wish_bug

When I exclude that gauntlet, I get "No builds found",
last_wish_bug2

I've tested this with other combat/seasonal mods including Taking Charge and Lucent Blade and the same issue occurs.

**Include devtools console logs**
If you can, open the devtools console (ctrl+shift+J on PC and cmd+option+J on Mac) and paste what you see in the window that pops up:

```
[loadout optimizer] found 200 stat mixes after processing 1692672 stat combinations in 5427.8249999915715 ms
log.ts:11 [loadout optimizer] useProcess: worker time 5472.185000020079ms
log.ts:11 [loadout optimizer] useProcess 5516.835000016727ms
log.ts:11 [loadout optimizer] found 0 stat mixes after processing 1639776 stat combinations in 4766.585000004852 ms
log.ts:11 [loadout optimizer] useProcess: worker time 4832.87000001315ms
log.ts:11 [loadout optimizer] useProcess 4855.414999998175ms
```

### PR DESCRIPTION
**Pull request by ryan-rushton**
This should be a fix to all the mod related issues we were having in the optimiser.

I am pretty sure this is a list of all the things done,

- Introduce two mod picker categories, other (for legacy and combat mods) and raid. Raid obviously will fit in legacy but as it has its own socket on new armour I had to split it out
- Fix the search when clicking on an item socket, there is a little funkyness there as when we click on a combat mod we dont want to show nightmare ect, so we detect that we clicked on a combat and filter out anything that doesnt fit in the combat style category plug hashes
- Remove all the season attributes hanging around on optimiser mods and armour
- Remove the specialty socket for plug category hashes as its not really useful now there is a bunch of overlap. We need to group these things manually now unfortunately.

The new mod sections.
image
image

Fixes #6192
Fixes #6195

### CODE REVIEW

File: src/locale/dim.json [hunk 1]
```diff
@@ -415,7 +418,6 @@
   "LoadoutBuilder": {
     "AssumeMasterwork": "Assume Masterworked Stats",
     "AssumeMasterworkDetailed": "Generate sets assuming all items are masterworked.",
-    "BadPerk": "A perk that \"slightly\" modifies isn't the best option.",
```
Comment (ryan-rushton) [inline src/locale/dim.json:418]: @delphiactual I seem to have i18n cleaning up a bunch of stuff. Do you want me to back these out?

Comment (delphiactual) [reply src/locale/dim.json:418]: nope go ahead and let them through

**Review by bhollis — APPROVED** [review]
This is a lot! It was pretty hard for me to understand what the changes were, but nothing looks egregious. It makes me think we should try and figure out how to make all of this more data-driven, either from the manifest (ideally) or computed d2ai data.

File: src/app/loadout-builder/filter/ModPicker.tsx [hunk 2]
```diff
@@ -51,6 +54,7 @@ interface ProvidedProps {
   lockedArmor2Mods: LockedArmor2ModMap;
   classType: DestinyClass;
   initialQuery?: string;
+  filterLegacy?: boolean;
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/filter/ModPicker.tsx:57]: Yup I will do another one and add a comment.

File: src/app/loadout-builder/filter/ModPicker.tsx [hunk 3]
```diff
@@ -240,22 +239,27 @@ function ModPicker({
...
-                perkDef &&
-                (regexp.test(perkDef.displayProperties.name) ||
-                  regexp.test(perkDef.displayProperties.description) ||
-                  regexp.test(perk.requirementDisplayString))
-              );
-            })
+            // TODO I am not thrilled about this filter legacy thing but it will stop
+            // legacy mods being shown when you click on a combat mod socket. Lets aim to
+            // get proper search in here at some point.
+            (!filterLegacy ||
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/filter/ModPicker.tsx:245]: No if filterLegacy is false it skips the check on combat compatible plugs. Basically it just skips this step and moves onto the normal search bit.

**Comment by ryan-rushton** [comment]
Yeah unfortunately it was one of those scenarios where fixing one thing just broke another and all the things accumulated.

**Comment by ryan-rushton** [comment]
Also regarding it being made more data driven, I am all for that. The mod assignment algorithm might be a bit hard to get that happening as its really touchy on a performance front. If you have some ideas though we could come up with something.

**Pull request merged**